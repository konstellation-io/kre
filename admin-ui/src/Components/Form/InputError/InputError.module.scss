@import 'Styles/mixins';
@import 'Styles/colors';

.errorMessage {
  @include input-footer($color-error);
}
