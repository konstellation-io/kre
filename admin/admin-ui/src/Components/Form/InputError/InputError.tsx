import React from 'react';
import cx from 'classnames';
import styles from './InputError.module.scss';

type Props = {
  message?: string;
};

function InputError({ message = '' }: Props) {
  return (
    <div
      className={cx(styles.errorMessage, { [styles.show]: message !== '' })}
      data-testid="error-message"
    >
      {message}
    </div>
  );
}

export default InputError;
