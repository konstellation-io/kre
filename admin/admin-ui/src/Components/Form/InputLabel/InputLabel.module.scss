@import 'Styles/mixins';
@import 'Styles/colors';

.label {
  @include font-caption;

  display: block;
  letter-spacing: 0.4px;
  color: font-color(grey);
  margin-bottom: 11px;
  min-height: 13px;

  transition: color ease 0.3s;

  &.hidden {
    color: transparent;
  }
}
