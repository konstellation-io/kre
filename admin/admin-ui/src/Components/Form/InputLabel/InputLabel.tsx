import React from 'react';
import cx from 'classnames';
import styles from './InputLabel.module.scss';

type Props = {
  text?: string;
  hidden?: boolean;
};

function InputLabel({ text = '', hidden = false }: Props) {
  return (
    <label
      className={cx(styles.label, { [styles.hidden]: hidden })}
      data-testid="label"
    >
      {text.toUpperCase()}
    </label>
  );
}

export default InputLabel;
