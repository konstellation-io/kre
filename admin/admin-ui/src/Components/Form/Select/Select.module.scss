@import 'Styles/mixins';
@import 'Styles/colors';
@import 'Styles/borders';
@import 'Styles/shadows';

$height-input: 30px;

.container {
  position: relative;

  &.white,
  &.white .input {
    color: white;
  }

  &.hasClearButton .input {
    padding-right: 30px;
  }
}

.inputContainer {
  position: relative;
}

.input {
  @include form-input;
  padding: 11px 40px 11px 16px;
  line-height: normal;
  white-space: nowrap;
  cursor: pointer;

  color: font-color(light);
  font-weight: 600;

  &.placeholder {
    color: font-color(lowlight);
  }

  &:after {
    @include arrow(white);

    position: absolute;
    width: 7px;
    height: 7px;
    transform: rotate(45deg);
    right: 14px;
    top: 14px;
    content: '';
    transition: transform ease 0.3s, top ease 0.3s;
  }

  &.opened:after {
    transform: rotate(45deg) scale(-1);
    top: 17px;
  }

  &.disabled {
    color: font-color(regular);
    background-color: palette(base, 400);
    pointer-events: none;
  }
}

.optionsContainer {
  @include shadow(3);

  position: absolute;
  background-color: $bg-shape-default;
  z-index: 1;
  top: 100%;
  right: 0;
  transition: max-height ease 0.3s, top ease 0.6s, overflow ease 0.6s;
  overflow: hidden;
  width: max-content;
  min-width: 100%;
  border-radius: 5px;

  &.opened {
    top: 100%;
    animation: show-overflow 0s ease-in 0.3s forwards;
  }
}

@keyframes show-overflow {
  to {
    overflow: auto;
  }
}

.defaultOption {
  height: 40px;
  padding: 11px 16px;
}

.optionElement {
  color: font-color(light);
  line-height: 16px;
  cursor: pointer;

  &.selected {
    background-color: $bg-shape-selected;
  }

  &.disabled {
    pointer-events: none;
    cursor: default;
  }

  &:hover {
    @include glow(3);
    background-color: $bg-shape-hover;
    color: font-color(dark);
  }
}

.container.light {
  .input {
    background-color: transparent;
    color: font-color(lowlight);
    border: 0;
  }
}
