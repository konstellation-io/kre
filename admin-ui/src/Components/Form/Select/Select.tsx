import React, { FC, useEffect, useRef, useState } from 'react';

import InputError from '../InputError/InputError';
import InputLabel from '../InputLabel/InputLabel';
import cx from 'classnames';
import { get } from 'lodash';
import styles from './Select.module.scss';
import useClickOutside from 'Hooks/useClickOutside';

const MAX_HEIGHT = 240;

export enum SelectTheme {
  DEFAULT = 'default',
  LIGHT = 'light'
}

export type CustomOptionProps = {
  label: string;
};

type Props = {
  onChange?: Function;
  label?: string;
  height?: number;
  error?: string;
  whiteColor?: boolean;
  shouldSort?: boolean;
  placeholder?: string;
  defaultOption?: string;
  options: string[];
  formSelectedOption?: string;
  valuesMapper?: { [key: string]: string };
  selectMainClass?: string;
  hideError?: boolean;
  className?: string;
  disabled?: boolean;
  theme?: SelectTheme;
  disabledOptions?: string[];
  CustomOptions?: { [key: string]: FC<CustomOptionProps> };
  showSelectAllOption?: boolean;
};

function Select({
  options,
  onChange = function() {},
  label = '',
  height = 40,
  error = '',
  whiteColor = false,
  shouldSort = true,
  disabled = false,
  defaultOption,
  placeholder = '',
  formSelectedOption,
  valuesMapper = {},
  selectMainClass = '',
  hideError = false,
  className = '',
  theme = SelectTheme.DEFAULT,
  CustomOptions = {},
  disabledOptions = [],
  showSelectAllOption = true
}: Props) {
  const optionsRef = useRef<HTMLDivElement>(null);
  const { addClickOutsideEvents, removeClickOutsideEvents } = useClickOutside({
    componentRef: optionsRef,
    action: closeOptions
  });
  const selectedOptionRef = useRef<HTMLDivElement>(null);
  const [selectedOption, setSelectedOption] = useState<
    string | undefined | null
  >(placeholder || defaultOption);
  const [optionsOpened, setOptionsOpened] = useState(false);

  useEffect(() => {
    setSelectedOption(formSelectedOption || placeholder || defaultOption);
  }, [
    options,
    defaultOption,
    selectedOption,
    setSelectedOption,
    formSelectedOption,
    placeholder
  ]);

  useEffect(() => {
    if (optionsOpened && selectedOptionRef.current !== null) {
      selectedOptionRef.current?.scrollIntoView({
        block: 'end'
      });
    }
  }, [optionsOpened]);

  function openOptions() {
    if (!optionsOpened) {
      addClickOutsideEvents();
      setOptionsOpened(true);
    }
  }

  function closeOptions() {
    removeClickOutsideEvents();
    setOptionsOpened(false);
  }

  function handleOnOptionCLick(value: string | null) {
    closeOptions();
    setSelectedOption(value);
    onChange(value);
  }

  function renderOption(option: string) {
    const label = get(valuesMapper, option, option);
    const CustomOption = CustomOptions[option];
    return CustomOption ? (
      <CustomOption label={label} />
    ) : (
      <div className={styles.defaultOption}>{label}</div>
    );
  }

  const optionList = (shouldSort ? [...options].sort() : [...options]).map(
    (option: string, idx: number) => (
      <div
        key={`${option}-${idx}`}
        className={cx(styles.optionElement, {
          [styles.selected]: option === selectedOption,
          [styles.disabled]: disabledOptions.includes(option)
        })}
        onClick={() => handleOnOptionCLick(option)}
        ref={option === formSelectedOption ? selectedOptionRef : null}
      >
        {renderOption(option)}
      </div>
    )
  );
  if (placeholder && showSelectAllOption) {
    optionList.unshift(
      <div
        key={`selectOption_empty`}
        className={cx(styles.optionElement, {
          [styles.selected]: selectedOption === placeholder
        })}
        onClick={() => handleOnOptionCLick(null)}
      >
        {renderOption('All')}
      </div>
    );
  }
  const optionsHeight = Math.min(optionList.length * 40, MAX_HEIGHT);

  return (
    <div
      className={cx(
        className,
        styles[theme],
        styles.container,
        selectMainClass,
        {
          [styles.white]: whiteColor
        }
      )}
    >
      {label && <InputLabel text={label} />}
      <div className={styles.inputContainer}>
        <div
          className={cx(styles.input, {
            [styles.error]: error !== '',
            [styles.opened]: optionsOpened,
            [styles.disabled]: disabled,
            [styles.placeholder]: placeholder === selectedOption
          })}
          style={{ height }}
          onClick={openOptions}
        >
          {get(valuesMapper, selectedOption || '', selectedOption)}
        </div>
        <div
          className={cx(styles.optionsContainer, {
            [styles.opened]: optionsOpened
          })}
          ref={optionsRef}
          style={{ maxHeight: optionsOpened ? optionsHeight : 0 }}
        >
          {optionsOpened && optionList}
        </div>
      </div>
      {!hideError && <InputError message={error} />}
    </div>
  );
}

export default Select;
