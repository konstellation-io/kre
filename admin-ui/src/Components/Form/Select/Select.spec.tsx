import React from 'react';
import Select from './Select';
import InputLabel from '../InputLabel/InputLabel';
import InputError from '../InputError/InputError';
import { shallow } from 'enzyme';

const OPTIONS = ['Option A', 'Option B', 'Option C'];

const mockOnChange = jest.fn();

describe('Select', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(
      <Select
        label="Select"
        placeholder="placeholder"
        options={OPTIONS}
        onChange={mockOnChange}
      />
    );
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('show right components', () => {
    expect(wrapper.find(InputLabel).prop('text')).toBe('Select');
    expect(wrapper.find(InputError).prop('message')).toBe('');
    expect(
      wrapper
        .find('.optionElement')
        .at(0)
        .text()
    ).toBe('All');
    expect(wrapper.exists('.placeholder')).toBeTruthy();
    expect(wrapper.exists('.opened')).toBeFalsy();
  });

  it('show error message', () => {
    wrapper.setProps({ error: 'Some error' });

    expect(wrapper.find(InputError).prop('message')).toBe('Some error');
  });

  it('handles events', () => {
    wrapper.find('.input').simulate('click');
    expect(wrapper.exists('.opened')).toBeTruthy();

    wrapper
      .find('.optionElement')
      .at(1)
      .simulate('click');

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(OPTIONS[0]);
    expect(wrapper.exists('.opened')).toBeFalsy();
  });
});
