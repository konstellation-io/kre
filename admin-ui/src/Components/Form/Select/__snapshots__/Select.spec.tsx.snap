// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select matches snapshot 1`] = `
<div
  className="default container"
>
  <InputLabel
    text="Select"
  />
  <div
    className="inputContainer"
  >
    <div
      className="input placeholder"
      onClick={[Function]}
      style={
        Object {
          "height": 40,
        }
      }
    >
      placeholder
    </div>
    <div
      className="optionsContainer"
      style={
        Object {
          "maxHeight": 0,
        }
      }
    >
      <div
        className="optionElement selected"
        key="selectOption_empty"
        onClick={[Function]}
      >
        <div
          className="defaultOption"
        >
          All
        </div>
      </div>
      <div
        className="optionElement"
        key="Option A-0"
        onClick={[Function]}
      >
        <div
          className="defaultOption"
        >
          Option A
        </div>
      </div>
      <div
        className="optionElement"
        key="Option B-1"
        onClick={[Function]}
      >
        <div
          className="defaultOption"
        >
          Option B
        </div>
      </div>
      <div
        className="optionElement"
        key="Option C-2"
        onClick={[Function]}
      >
        <div
          className="defaultOption"
        >
          Option C
        </div>
      </div>
    </div>
  </div>
  <InputError
    message=""
  />
</div>
`;
